modin/backends/pandas/query_compiler.py

```diff
@@ -2576,34 +2576,71 @@ def groupby_agg(
         if callable(agg_func):
             agg_func = wrap_udf_function(agg_func)
 
-        if is_multi_by:
-            return super().groupby_agg(
-                by=by,
-                is_multi_by=is_multi_by,
-                axis=axis,
-                agg_func=agg_func,
-                agg_args=agg_args,
-                agg_kwargs=agg_kwargs,
-                groupby_kwargs=groupby_kwargs,
-                drop=drop,
-            )
-
-        by = by.to_pandas().squeeze() if isinstance(by, type(self)) else by
-
         # since we're going to modify `groupby_kwargs` dict in a `groupby_agg_builder`,
         # we want to copy it to not propagate these changes into source dict, in case
         # of unsuccessful end of function
         groupby_kwargs = groupby_kwargs.copy()
 
         as_index = groupby_kwargs.get("as_index", True)
 
-        def groupby_agg_builder(df):
+        if isinstance(by, type(self)):
+            internal_by = by.columns
+            by = [by]
+        else:
+            if not isinstance(by, list):
+                by = [by]
+            internal_by = [o for o in by if isinstance(o, str) and o in self.columns]
+            internal_qc = (
+                [self.getitem_column_array(internal_by)] if len(internal_by) else []
+            )
+
+            by = internal_qc + by[len(internal_by) :]
+
+        broadcastable_by = [o._modin_frame for o in by if isinstance(o, type(self))]
```

Comment:
This is a good idea, but in this particular case, this won't give us any benefit.

numpy array will not fit us to store Series and DataFrame objects, because they implement `__array_interface__` and so numpy will grab their data as a pure array, without preserving metadata such as names:
```python
>>> np.array([pandas.Series([1, 2, 3])], dtype=object)
array([[1, 2, 3]], dtype=object)
```

So we want to have `by` as a list inside of `groupby_agg_builder`, and we also want to [concatenate](https://github.com/dchigarev/modin/blob/5a6c93dad9a658c8f7a09eed020ceba5df7b41a8/modin/backends/pandas/query_compiler.py#L2641) `by` list with `not_broadcastable_by`, so it also have to be a python list.

So in case of using mask instead of 2 for loops via 'by' we will get 3:
1. iterate trough `by` to create a mask
2. create a numpy array from `by`, to be able to grab elements from it via mask
3. convert `not_broadcastable_by` back to list again, to be able to append it at `by` inside `groupby_agg_builder`

Reply:
I meant something like that:
```python
bys = []
for o in by:
    if isinstance(o, type(self)):
        bys.append([o._modin_frame, 1])
    else:
        bys.append([o, 0])
np_bys = np.array(bys)
masked_np_bys = np.ma.array(bys[:, 0], mask=bys[:,1])
# and then work with this array, but it looks complicated too
```
Let's leave current approach via lists.